sql: allow RETURNING clause to access all columns

RaduBerinde opened the pull request:
The RETURNING clause can only access columns that are being updated or inserted.
This is fixed by always providing all columns to ReturningHelper. For update we
already have values for all columns in rowVals; for insert we fill in the
provided values and let the other values be NULL.

Fixes #4368.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/5044)

<!-- Reviewable:end -->

petermattis commented:
Review status: 0 of 4 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[sql/insert.go, line 231 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5044#-KCTOvkTq-q7IyEvJ--u:-KCTOvkUsuYaDW9SOeBC:-1793290945) ([raw file](https://github.com/cockroachdb/cockroach/blob/ba674baeafd1387aa071a8710301144fc793c452/sql/insert.go#L231)):*
I've been staring at this code for a while. How is `retVals` different from `rowVals`? Seems like `rowVals` has an entry indexed by column index for every column in the table. Isn't that exactly what `retVals` contains.

---

*[sql/update.go, line 314 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5044#-KCTPaKD5XPin_m7UMk7:-KCTPaKD5XPin_m7UMk8:779314762) ([raw file](https://github.com/cockroachdb/cockroach/blob/ba674baeafd1387aa071a8710301144fc793c452/sql/update.go#L314)):*
Don't we want `RETURNING` to return the newly updated value, not the previous value? Guess I'm confused as to how this is working.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5044)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 4 of 4 files at r1.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[sql/insert.go, line 224 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5044#-KCTVL943SO3WG49JCmb:-KCTVL943SO3WG49JCmc:948298741) ([raw file](https://github.com/cockroachdb/cockroach/blob/ba674baeafd1387aa071a8710301144fc793c452/sql/insert.go#L224)):*
i think this needs a comment (what you have in the commit message is sufficient)

---

*[sql/testdata/insert, line 339 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5044#-KCTUde2IZ_uBJv5dYpc:-KCTUde2IZ_uBJv5dYpd:598525202) ([raw file](https://github.com/cockroachdb/cockroach/blob/ba674baeafd1387aa071a8710301144fc793c452/sql/testdata/insert#L339)):*
can you give `b` a default so that this test returns an interesting value?

---

*[sql/testdata/update, line 193 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5044#-KCTUxnM1ttRlIGjjG0E:-KCTUxnM1ttRlIGjjG0F:589168924) ([raw file](https://github.com/cockroachdb/cockroach/blob/ba674baeafd1387aa071a8710301144fc793c452/sql/testdata/update#L193)):*
do we have a test that does `RETURNING *, 4`?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5044)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 1 of 4 files reviewed at latest revision, 5 unresolved discussions.

---

*[sql/insert.go, line 231 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5044#-KCTOvkTq-q7IyEvJ--u:-KCTWq__KChUiw2fm49e:-1844155897) ([raw file](https://github.com/cockroachdb/cockroach/blob/ba674baeafd1387aa071a8710301144fc793c452/sql/insert.go#L231)):*
There are some cases where `rowVals` doesn't have a value for each column (e.g. when doing `INSERT INTO return (a)` it will only have a value for `a`).

---

*[sql/update.go, line 314 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5044#-KCTPaKD5XPin_m7UMk7:-KCTXe_5zhYIfUl4wTTI:184033298) ([raw file](https://github.com/cockroachdb/cockroach/blob/ba674baeafd1387aa071a8710301144fc793c452/sql/update.go#L314)):*
We update them above (grep for "Update the row values"). Adding a comment.

---

*[sql/testdata/insert, line 339 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5044#-KCTUde2IZ_uBJv5dYpc:-KCTYQN9RBW9b9hg1lAr:230058591) ([raw file](https://github.com/cockroachdb/cockroach/blob/ba674baeafd1387aa071a8710301144fc793c452/sql/testdata/insert#L339)):*
I think that would have worked without this change (we have to insert columns with defaults so `b` should have been accessible). There are a bunch of existing tests around default values.

---

*[sql/testdata/update, line 193 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5044#-KCTUxnM1ttRlIGjjG0E:-KCTYx6RDBLHlQv9V1bu:-1629460504) ([raw file](https://github.com/cockroachdb/cockroach/blob/ba674baeafd1387aa071a8710301144fc793c452/sql/testdata/update#L193)):*
Adding.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5044)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[sql/update.go, line 314 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5044#-KCTcIg3HyrSzHRf_DI4:-KCTcIg4T0sUNRRFhzxF:-205248540) ([raw file](https://github.com/cockroachdb/cockroach/blob/75503c0948324e266986b9fed723f47e5fdf9023/sql/update.go#L314)):*
nit: the `0` is not necessary (but you can leave it if you prefer)

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5044)*
<!-- Sent from Reviewable.io -->

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> but let @petermattis sign off

---

Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5044#-:-KCTcX75_3PL_VNwF0wg:-881263185)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 3 of 4 files reviewed at latest revision, 3 unresolved discussions.

---

*[sql/update.go, line 314 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5044#-KCTcIg3HyrSzHRf_DI4:-KCTlVTXfrNj6oB5vy6A:-624251961) ([raw file](https://github.com/cockroachdb/cockroach/blob/75503c0948324e266986b9fed723f47e5fdf9023/sql/update.go#L314)):*
Ah right, fixed. TFTR!

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5044)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5044)*
<!-- Sent from Reviewable.io -->

mjibson commented:
LGTM

---

Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[sql/update.go, line 314 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5044#-KCTyn_tfsdaVmNMgXyC:-KCTyn_upXXReVNkVKJY:998581664) ([raw file](https://github.com/cockroachdb/cockroach/blob/1b0a3836d611b01d777f6cc27bdf90283b2c047b/sql/update.go#L314)):*
missing a closing )

---

*[sql/testdata/insert, line 372 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5044#-KCTwNLG0on4X6fqW88b:-KCTwNLHoho_If3Af6bG:-1089556699) ([raw file](https://github.com/cockroachdb/cockroach/blob/1b0a3836d611b01d777f6cc27bdf90283b2c047b/sql/testdata/insert#L372)):*
Could we add a test to make sure default values can be seen by the returning expression:

query III
INSERT into return (b) VALUES (1) RETURNING *, a + 1
----
3 1 4

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5044#-:-KCTzH2d014U4lgplpu_:676834233)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 2 of 4 files reviewed at latest revision, 4 unresolved discussions.

---

*[sql/testdata/insert, line 372 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5044#-KCTwNLG0on4X6fqW88b:-KCVXqfbJl8lQJCeFxj4:1478236514) ([raw file](https://github.com/cockroachdb/cockroach/blob/1b0a3836d611b01d777f6cc27bdf90283b2c047b/sql/testdata/insert#L372)):*
Added, TFTR!

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5044)*
<!-- Sent from Reviewable.io -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 2 of 4 files reviewed at latest revision, 5 unresolved discussions.

---

*[sql/insert.go, line 231 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5044#-KCTOvkTq-q7IyEvJ--u:-KCVicoE-CHm6dWtxi_V:413652410) ([raw file](https://github.com/cockroachdb/cockroach/blob/ba674baeafd1387aa071a8710301144fc793c452/sql/insert.go#L231)):*
Ok, got it. Rather than using a map for `colIDToRetIndex`, you could use a `rowValsIndexToRetIndex` slice that is indexed by `i`. Somewhat more irritating to initialize, though.

---

*[sql/insert.go, line 141 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5044#-KCVjK3oDEDKFMXceQ5n:-KCVjK3p5nWAKzNeJkPo:-345689239) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c1141f333387c455fa956af70f4190b3ceda044/sql/insert.go#L141)):*
Can we avoid doing this work if we're not returning anything?

---

*[sql/update.go, line 314 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5044#-KCTPaKD5XPin_m7UMk7:-KCVj_LrQLQM6zO7Te2x:1230478415) ([raw file](https://github.com/cockroachdb/cockroach/blob/ba674baeafd1387aa071a8710301144fc793c452/sql/update.go#L314)):*
Ah, got it.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5044#-:-KCVjfQ2W1gG_4BaoDnP:1417088721)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 1 of 4 files reviewed at latest revision, 3 unresolved discussions.

---

*[sql/insert.go, line 141 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5044#-KCVjK3oDEDKFMXceQ5n:-KCWFWb_pK4ASpwkgDah:-719267080) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c1141f333387c455fa956af70f4190b3ceda044/sql/insert.go#L141)):*
Updated.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5044)*
<!-- Sent from Reviewable.io -->

RaduBerinde committed:
sql: allow RETURNING clause to access all columns

The RETURNING clause can only access columns that are being updated or inserted.
This is fixed by always providing all columns to ReturningHelper. For update we
already have values for all columns in rowVals; for insert we fill in the
provided values and let the other values be NULL.

Fixes #4368.
